### DISCUSSION TITLE
Problem with connection

### DISCUSSION THREAD

**Discussion by MartinGruenewald** [robert-rc2i/ptq-client-webapp#28] [Q&A] [answered]
I'm running Pianoteq on an Intel NUC under Windows 11.
I did everything as described, and when I type http://ip.of.my.nuc:8081/jsonrpc in my browser, I see a page which explains the commands.
But with "http://ip.of.my.nuc:8081/jsonrpc" in your web remote, I always get the response "Network error".
I've installed Node, too - from the link.
Do you have any ideas, what I could try?
Regards
Martin

**Comment by robert-rc2i** [comment]
Hi Martin,

I am assuming that ip.of.my.nuc is a placeholder right? Also, when you click on the menu button (Top left), can you confirm that you see your URL in the `Pianoteq full API address` field? If not, make sure that when you enter your URL that you click on the blue right arrow to submit the info.

Otherwise, for me to be able to help, I would require the following:

1. Provides the content of the browser's console log while you run the web application
    1. Safari : Under the menu Develop -> Show web inspector.
    2. Chrome :  [See how here](https://developer.chrome.com/docs/devtools/open/#chrome)
2. Then select the Console tab and cut and paste its content
3. In the storage tab, please confirm that you have an entry "ptqUrl" with the value you entered in the app.

Regards

**Reply by MartinGruenewald** [reply]
Hi Robert,
thanks for your fast reply!

in the storage tab I see:
ptqURL=http://192.168.10.69:8081/jsonrpc 
which is the correct URL.

And this is reported in the console:

Access to fetch at 'http://192.168.10.69:8081/jsonrpc' from origin 'http://ptq-web-client.s3-website-us-east-1.amazonaws.com' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `private`.
main.0190233b.js:2          POST http://192.168.10.69:8081/jsonrpc net::ERR_FAILED
(anonymous) @ main.0190233b.js:2
c @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
i @ main.0190233b.js:2
l @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
value @ main.0190233b.js:2
value @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
c @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
i @ main.0190233b.js:2
l @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
Et @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
c @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
i @ main.0190233b.js:2
l @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
Dt @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
c @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
i @ main.0190233b.js:2
l @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
Lt @ main.0190233b.js:2
onClick @ main.0190233b.js:2
Ve @ main.0190233b.js:2
Ge @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
Nr @ main.0190233b.js:2
Or @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
De @ main.0190233b.js:2
(anonymous) @ main.0190233b.js:2
Mr @ main.0190233b.js:2
Jt @ main.0190233b.js:2
Zt @ main.0190233b.js:2
t.unstable_runWithPriority @ main.0190233b.js:2
qa @ main.0190233b.js:2
Le @ main.0190233b.js:2
Yt @ main.0190233b.js:2
Show 24 more frames
main.0190233b.js:2 Network: Error e {code: -1, message: 'TypeError: Failed to fetch', time: 1660753457309}

**Reply by robert-rc2i** [reply]
Hi Martin,

Thanks for the additional info.  That is very strange, as Pianoteq does not impose any CORS restriction.  Pianoteq is supposed to return this header when calling `/jsonrpc`

* Access-Control-Allow-Origin: *

The above authorize request from all origins.

OK, I'll need more information about your setup.

* Browser version
* OS where the browser is being used
* Is there anything in between the browser and Pianoteq?  I.e. a proxy?
* Lastly, can you provide the output of the network tab, more specifically for the `jsonrpc` ones. with the content of:
    * Preview 
    * Headers

Sordess

**Reply by robert-rc2i** [reply]
One last thing.  

When you launch the PTQ-Client-Webapp.  Are you launching it with HTTPS?  If so, DO NOT.  Pianoteq is serving HTTP request only.  We cannot mix HTTPS with HTTP fetch calls.  Under that scenario you will get a CORS message stating what you mention above.

Sordess

**Reply by robert-rc2i** [reply]
Humm, after reading a bit more your log capture, I think I know why

1. You are not using HTTPS, which is good!
2. It appears you are using Chrome
3. Chrome decided to implement a new security protocol for external to local calls.  Which is actually a good thing for security reason.  But this change will definitely require a new directive from Pianoteq to make it work.
    1. [Private network access](https://developer.chrome.com/blog/private-network-access-preflight/)

I'll have to ask Mozart to add this required directive, otherwise that public site won't work anymore.

In the mean time try with another browser to see if it works.  Otherwise, your best option would be for you to host this site on your device where you run Pianoteq.

Sordess

**Comment by robert-rc2i** [comment, marked as answer]
## Probable cause
This is caused by a new version of Chrome, which is implementing a new security option that forbids external sites to make local network fetch calls

See this site here for more information: https://developer.chrome.com/blog/private-network-access-preflight/

## Resolution
There is nothing our application can do to avoid this.  This security measure is imposed by the browser.  It is actually good thing as it will make for safer browsing, but will negatively impact the ability for me to host this site publicly.

## Fallback options

1. Change browser, but it will eventually fail on these browsers as well when they also implement this security specification.  Tested browsers are
    1. Safari
    2. Firefox
2. Do not use the public site and host the application locally.  See the readme.md for more detail how.

**Reply by MartinGruenewald** [reply]
Thank you so much, Robert.
I've tried Safari on my iPad, and it works. I'll never use it, but wouldn't mind to do it to use your your app.
On my Samsung phone, it doesn't work with the Samsung browser. But I think the iPad is the best device anyway for this.
Thanks a lot,
Martin

**Reply by robert-rc2i** [reply]
Glad it worked out and it also help me learn a new specification that I was not aware!  :/ 

As for using it on your Samsung phone, it will only work if you host the web application on your network.  That way, all the calls will originate from your network, which is allowed by the spec.

Cheers
